apollo-router/src/plugins/diagnostics/memory/mod.rs

```diff
@@ -0,0 +1,386 @@
+                        "status": "stopped",
+                        "message": "Memory profiling deactivated"
+                    }),
+                ),
+                Err(e) => (
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                    json!({
+                        "status": "error",
+                        "message": format!("Failed to stop profiling: {}", e)
+                    }),
+                ),
+            }
+        };
+
+        // On non-Linux platforms, return not supported
+        #[cfg(not(target_os = "linux"))]
+        let (status_code, response) = (
+            StatusCode::NOT_IMPLEMENTED,
+            json!({
+                "status": "not_supported",
+                "message": format!("Memory profiling not supported on platform: {}", std::env::consts::OS),
+                "platform": std::env::consts::OS
+            }),
+        );
+
+        Response::http_response_builder()
+            .response(
+                http::Response::builder()
+                    .status(status_code)
+                    .header("content-type", "application/json")
+                    .body(body::from_bytes(serde_json::to_vec(&response)?))?,
+            )
+            .context(request.context)
+            .build()
+    }
+
+    /// Handle POST /diagnostics/memory/dump
+    pub(super) async fn handle_dump(&self, request: Request) -> Result<Response, BoxError> {
+        tracing::info!("Memory dump requested");
+
+        // On Linux, perform actual heap dump
+        #[cfg(target_os = "linux")]
+        let (status_code, response) = {
+            let base_output_directory = self.output_directory.clone();
+
+            match tokio::task::spawn_blocking(move || -> Result<String, MemoryError> {
+                // Generate timestamp for the dump file
+                let timestamp = SystemTime::now()
+                    .duration_since(UNIX_EPOCH)
+                    .map_err(|e| MemoryError::SystemCall(e.to_string()))?
+                    .as_secs();
+
+                // Create memory subdirectory structure to mirror archive
+                let base_path = Path::new(&base_output_directory);
+                let memory_path = base_path.join("memory");
+                fs::create_dir_all(&memory_path).map_err(|e| {
+                    MemoryError::SystemCall(format!(
+                        "Failed to create memory dump directory {}: {}",
+                        memory_path.display(),
+                        e
+                    ))
+                })?;
+
+                let dump_path = memory_path
+                    .join(format!("router_heap_dump_{}.prof", timestamp))
+                    .to_string_lossy()
+                    .to_string();
+
+                // Create CString for the dump path
+                let value = CString::new(dump_path.clone()).map_err(|e| {
+                    MemoryError::SystemCall(format!("Failed to create CString: {}", e))
+                })?;
```

Comment:
Can't we just unwrap here? Wouldn't this only happen if we couldn't allocate?

Reply:
Not sure, let's figure this out.